### PR TITLE
UI for Magic Link flow

### PR DESCRIPTION
**Pull request by frankie567**
Here is the basic flow for Magic Link authentication!

1. Users input their email on the login page
2. They receive an email with the link
3. They click on the link and are automatically authenticated 🎉 

## Demo time!

https://github.com/polarsource/polar/assets/1144727/f65e7c02-8025-499e-a9b4-066ca038a456

### Light mode preview

Capture d’écran 2023-09-13 à 15 01 37

Capture d’écran 2023-09-13 à 15 01 48

## Changes to the existing code

* `/users/me` now output the list of `oauth_accounts` (only the platform, account ID and email, of course 🔒). It'll help us to detect if the users is authenticated through GitHub or not
* We use this new capability to hide the **Recommended** feed if not connected to GitHub. *Could be the opportunity to have a GitHub upsell banner instead*
* Hide the avatar if we don't have it

### CODE REVIEW

**Comment by birkjernstrom** [comment]
Super nice to see this and with the enhanced email template too. It's going to be really great to have this feature in-place, nicely done @frankie567!

I noticed there is a temporary error message displayed upon clicking the link: "Oh no, magic link...". I recognize this issue for GitHub too, I've received it once or twice. Only a moments error before success status. I bet there is some state issue that's causing this momentary issue. Can you investigate to see what's causing it? 🙏

File: clients/apps/web/src/app/login/magic-link/request/page.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    <div className="flex h-screen w-full grow items-center justify-center bg-[#FEFDF9] dark:bg-gray-950">
+      <div id="polar-bg-gradient"></div>
+      <div className="flex w-80 flex-col items-center">
+        <LogoType70 className="mb-6 h-10" />
+        <div className="text-center text-gray-500 dark:text-gray-400">
+          Email with sign in link sent to{' '}
+          <span className="font-bold">{email}</span>.
+        </div>
+        <div className="text-center text-gray-500 dark:text-gray-400">
+          If no account exists, it will be created.
```
Comment (birkjernstrom) [inline clients/apps/web/src/app/login/magic-link/request/page.tsx]: Let's change this to: "An account will be created unless one already exists."

**Comment by frankie567** [comment]
> I noticed there is a temporary error message displayed upon clicking the link: "Oh no, magic link...". I recognize this issue for GitHub too, I've received it once or twice. Only a moments error before success status. I bet there is some state issue that's causing this momentary issue. Can you investigate to see what's causing it? 🙏

Yeah, I saw it too! Nasty React re-render involved here. I've added a fix https://github.com/polarsource/polar/pull/1157/commits/2ced62b59a60aab3cec0cf432d8d83019cf749db (basically by breaking React hook rules ☺️) that seems to solve the issue. I can backport it for the GitHub callback too.

**Comment by zegl** [comment]
For some reason there is a single 0 rendered on this page. Is it related to your changes, or was it already there?

Screenshot 2023-09-13 at 17 17 36

**Comment by frankie567** [comment]
> For some reason there is a single 0 rendered on this page. Is it related to your changes, or was it already there?

I've it in prod, so yes, seemed to be already there!

Capture d’écran 2023-09-13 à 17 21 39

File: clients/apps/web/src/app/login/magic-link/authenticate/page.tsx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+export default function Page({
+  searchParams,
+}: {
+  searchParams: { [key: string]: string | string[] | undefined }
+}) {
+  const token = searchParams.token as string | undefined
+  const router = useRouter()
+  const session = useAuth()
+  const [error, setError] = useState<string | null>(null)
+  let loading = false
```
Comment (zegl) [inline clients/apps/web/src/app/login/magic-link/authenticate/page.tsx]: This value will reset on every render. Are you sure that you don't want to use a ref here?

**Comment by zegl** [comment]
> > For some reason there is a single 0 rendered on this page. Is it related to your changes, or was it already there?
> 
> I've it in prod, so yes, seemed to be already there!

Ok! I extracted it to a new issue to follow up #1160

File: clients/apps/web/src/app/login/magic-link/authenticate/page.tsx (hunk 2 above)
Comment (frankie567) [reply clients/apps/web/src/app/login/magic-link/authenticate/page.tsx]: Oh, right! Yes, much more cleaner this way! Never thought of using `useRef` for plain values.

**Pull request merged**